File: Sources/Scratchpad/Drawing/DrawingStoreFactory.swift
/// Creates and retains one `DrawingStore` per display, since v0.1 shows the
/// overlay on every connected monitor simultaneously and each screen's
/// drawing is independent (see DisplayIdentifier).
final class DrawingStoreFactory {
    private let repository: DrawingRepository
    private var storesByDisplay: [DisplayIdentifier: DrawingStore] = [:]

    init(repository: DrawingRepository) {
        self.repository = repository
    }

    func store(for display: DisplayIdentifier) -> DrawingStore {
        if let existing = storesByDisplay[display] {
            return existing
        }
        let store = DrawingStore(display: display, repository: repository)
        storesByDisplay[display] = store
        return store
    }

    func removeStore(for display: DisplayIdentifier) {
        storesByDisplay.removeValue(forKey: display)
    }

    var allStores: [DrawingStore] {
        Array(storesByDisplay.values)
    }
}

File: Windows/Scratchpad/Drawing/DrawingStoreFactory.cs
using Scratchpad.Overlay;
using Scratchpad.Persistence;

namespace Scratchpad.Drawing;

/// Creates and retains one `DrawingStore` per display, since v0.1 shows the
/// overlay on every connected monitor simultaneously and each screen's
/// drawing is independent (see `DisplayIdentifier`). Mirrors the macOS
/// app's `DrawingStoreFactory`.
public sealed class DrawingStoreFactory
{
    private readonly IDrawingRepository _repository;
    private readonly Dictionary<DisplayIdentifier, DrawingStore> _storesByDisplay = new();

    public DrawingStoreFactory(IDrawingRepository repository)
    {
        _repository = repository;
    }

    public DrawingStore Store(DisplayIdentifier display)
    {
        if (_storesByDisplay.TryGetValue(display, out var existing)) return existing;
        var store = new DrawingStore(display, _repository);
        _storesByDisplay[display] = store;
        return store;
    }

    public void RemoveStore(DisplayIdentifier display) => _storesByDisplay.Remove(display);

    public IReadOnlyList<DrawingStore> AllStores => _storesByDisplay.Values.ToList();
}
